internal/overlord/logstate/pumper.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	logReaderSize = 4 * 1024
+)
+
+// pumper is responsible for "pumping" the logs from the service ringbuffer
+// through to the log target client.
+type pumper struct {
+	// Metadata
+	service string
+	target  plan.LogTarget
+
+	iterator servicelog.Iterator
+	parser   *servicelog.Parser
+	client   logClient
+	stopCh   chan os.Signal
+}
+
+func newPumper(svc string, it servicelog.Iterator, tgt plan.LogTarget) (*pumper, error) {
+	client, err := newLogClient(svc, tgt)
+	if err != nil {
+		return nil, err
+	}
+
+	stopCh := make(chan os.Signal, 1)
+	// Notify on SIGINT so we can flush the buffer
+	signal.Notify(stopCh, unix.SIGINT)
+
+	return &pumper{
+		service:  svc,
+		target:   tgt,
+		iterator: it,
+		parser:   servicelog.NewParser(it, logReaderSize),
+		client:   client,
+		stopCh:   stopCh,
+	}, nil
+}
+
+// Pump continually pulls logs from the iterator and pushes them to the target.
+func (p *pumper) Pump() {
+	for {
+		// Check if Stop has been called
+		select {
+		case <-p.stopCh:
+			// force flush
+			p.client.Flush()
+			return
+		default:
+		}
+
+		// Let client flush if it's ready
+		p.client.MaybeFlush()
```

Comment:
I am not a fan of MaybeFlush() wording or what it's trying to do here. Specifically it's a function that is going to get called at ad nauseam for no reason and depending on the log back end it could be a very expensive call.

I think it's much more practical to have the pumper talk in real absolute terms. i.e I want you to flush now because I have more context then you do.

We can then move the time delta checker in loki back up to the pumper where I think it should live and we also reduce the interface size of log back end.

Reply:
ok, we should talk about the best way to divide the concerns here